### PR TITLE
#17751 Reserved folder names validation only for root-level folders

### PR DESCRIPTION
**Pull request by dsilvam**

### CODE REVIEW

File: dotCMS/src/main/java/com/dotmarketing/portlets/browser/ajax/BrowserAjax.java [hunk 1]
```diff
@@ -782,7 +783,6 @@ public boolean copyFolder ( String inode, String newFolder ) throws Exception {
      * @return Confirmation message
      * @throws Exception
      */
-    @WrapInTransaction
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotmarketing/portlets/browser/ajax/BrowserAjax.java]: why delete this annotation?

File: dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderFactoryImpl.java [hunk 2]
```diff
@@ -1208,12 +1223,16 @@ protected Identifier createIdentifierForFolder(Folder folder, String parentPath)
...
 	@Override
 	protected void save(Folder folderInode) throws DotDataException {
+		validateFolderName(folderInode);
+
 		HibernateUtil.getSession().clear();
 		HibernateUtil.saveOrUpdate(folderInode);
 	}
 
 	@Override
 	protected void save(Folder folderInode, String existingId) throws DotDataException {
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderFactoryImpl.java]: why do we have two save methods?,I think this method should call the other one

File: dotCMS/src/main/java/com/dotmarketing/portlets/browser/ajax/BrowserAjax.java [hunk 3]
```diff
@@ -800,7 +800,6 @@ public String moveFolder (final String folderId, final String newFolder) throws
             	return errorString;
 			}
         } catch (Exception e) {
-
         	Logger.error(this, e.getMessage(), e);
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/browser/ajax/BrowserAjax.java]: Please include also the folderId and newFolder values for logging

File: dotCMS/src/main/java/com/dotmarketing/portlets/browser/ajax/BrowserAjax.java (hunk 1 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotmarketing/portlets/browser/ajax/BrowserAjax.java]: None of the methods in this Ajax class have this annotation and this was actually causing an issue of not letting the API handle the rolling back in case of an exception at API level, thus leaving the system in a bad state.

File: dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderAPIImpl.java [hunk 4]
```diff
@@ -159,8 +147,10 @@ public boolean renameFolder(final Folder folder, final String newName,
 			Identifier folderId = APILocator.getIdentifierAPI().find(folder);
 			CacheLocator.getNavToolCache().removeNavByPath(folderId.getHostId(), folderId.getParentPath());
 			return renamed;
+		} catch (InvalidFolderNameException e) {
+			Logger.error(FolderAPIImpl.class, e.getMessage());
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderAPIImpl.java]: Not part of the PR, but can you include more useful logging data here please?

File: dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderFactoryImpl.java (hunk 2 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderFactoryImpl.java]: This is old code, not part of the change. I will check it out anyways to see if it can be refactored

File: dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderAPIImpl.java [hunk 5]
```diff
@@ -159,8 +147,10 @@ public boolean renameFolder(final Folder folder, final String newName,
...
+			throw e;
 		} catch (Exception e) {
-
 			throw new DotDataException(e.getMessage(),e);
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderAPIImpl.java]: And here?

File: dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderFactoryImpl.java [hunk 6]
```diff
@@ -1232,4 +1251,13 @@ protected void save(Folder folderInode, String existingId) throws DotDataExcepti
 			HibernateUtil.saveWithPrimaryKey(folderInode, existingId);
 		}
 	}
+
+	public void validateFolderName(final Folder folder) throws DotDataException {
+		if (UtilMethods.isSet(folder.getParentPermissionable())
+				&& folder.getParentPermissionable() instanceof Host
+				&& UtilMethods.isSet(folder.getName())
+				&& reservedFolderNames.contains(folder.getName().toUpperCase())) {
+			throw new InvalidFolderNameException("Folder can't be saved. You entered a reserved folder name");
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderFactoryImpl.java]: Include the name of the invalid folder in the error message.

File: dotCMS/src/main/java/com/dotmarketing/webdav/DotWebdavHelper.java [hunk 7]
```diff
@@ -994,9 +994,6 @@ public Folder createFolder(String folderUri, User user) throws IOException, DotD
 				throw new IOException(e.getMessage(),e);
 			}
 		} else {
-			if (host != null && InodeUtils.isSet(host.getInode())) {
-				folderAPI.validateFolderName(path.substring(1));
-			}
 			try {
 				hasPermission = perAPI.doesUserHavePermission(host, PERMISSION_CAN_ADD_CHILDREN, user, false);
 			} catch (DotDataException e) {
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/webdav/DotWebdavHelper.java]: Not part of the PR either, but can you include the folderUri in the error message please?

File: dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderFactoryImpl.java (hunk 2 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderFactoryImpl.java]: I don't see a way to refactor in an easy/safe way. Leaving it like it is.

**Review by jcastro-dotcms — CHANGES_REQUESTED** [review]
Requesting just a few improvements in error messages for easier troubleshooting.

**Review by freddyucv — APPROVED** [review]

File: dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderAPIImpl.java (hunk 5 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderAPIImpl.java]: Done

**Comment by dsilvam** [comment]
@jcastro-dotcms requested changes were included.

**Review by jcastro-dotcms — APPROVED** [review]

File: dotCMS/src/main/java/com/dotmarketing/portlets/browser/ajax/BrowserAjax.java (hunk 1 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/portlets/browser/ajax/BrowserAjax.java]: it is calling just one API, it would be ok

**Review by jdotcms — APPROVED** [review]
